File: demo-apps/android/app/src/main/kotlin/com/musicmuni/voxatrace/demo/sections/metronome/viewmodel/MetronomeViewModel.kt
package com.musicmuni.voxatrace.demo.sections.metronome.viewmodel

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.musicmuni.voxatrace.sonix.SonixMetronome
import io.github.aakira.napier.Napier
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File
import java.io.FileOutputStream

/**
 * ViewModel for metronome using SonixMetronome.
 */
class MetronomeViewModel : ViewModel() {

    // MARK: - Published State

    private val _isRunning = MutableStateFlow(false)
    val isRunning: StateFlow<Boolean> = _isRunning.asStateFlow()

    private val _bpm = MutableStateFlow(120f)
    val bpm: StateFlow<Float> = _bpm.asStateFlow()

    private val _beatsPerCycle = MutableStateFlow(4)
    val beatsPerCycle: StateFlow<Int> = _beatsPerCycle.asStateFlow()

    private val _currentBeat = MutableStateFlow(0)
    val currentBeat: StateFlow<Int> = _currentBeat.asStateFlow()

    private val _volume = MutableStateFlow(0.8f)
    val volume: StateFlow<Float> = _volume.asStateFlow()

    private val _status = MutableStateFlow("Ready")
    val status: StateFlow<String> = _status.asStateFlow()

    private val _isInitialized = MutableStateFlow(false)
    val isInitialized: StateFlow<Boolean> = _isInitialized.asStateFlow()

    // MARK: - Private State

    private var metronome: SonixMetronome? = null
    private var samaSamplePath: String? = null
    private var beatSamplePath: String? = null
    private val observerJobs = mutableListOf<Job>()

    // MARK: - Actions

    fun initialize(context: Context) {
        if (_isInitialized.value) return

        viewModelScope.launch {
            try {
                _status.value = "Initializing..."

                // Copy files on IO dispatcher
                val (samaPath, beatPath) = withContext(Dispatchers.IO) {
                    val samaFile = copyAssetToFile(context, "sama_click.wav")
                    val beatFile = copyAssetToFile(context, "beat_click.wav")
                    Pair(samaFile.absolutePath, beatFile.absolutePath)
                }

                samaSamplePath = samaPath
                beatSamplePath = beatPath

                createMetronome(samaPath, beatPath)
            } catch (e: Exception) {
                Napier.e("Metronome init failed", e)
                _status.value = "Error: ${e.message}"
            }
        }
    }

    private fun createMetronome(samaPath: String, beatPath: String) {
        // Clean up existing
        observerJobs.forEach { it.cancel() }
        observerJobs.clear()
        metronome?.release()

        val newMetronome = SonixMetronome.Builder()
            .samaSamplePath(samaPath)
            .beatSamplePath(beatPath)
            .bpm(_bpm.value)
            .beatsPerCycle(_beatsPerCycle.value)
            .volume(_volume.value)
            .onBeat { beat ->
                Napier.d("Beat $beat")
            }
            .onError { error ->
                Napier.e("Metronome error: $error")
                _status.value = "Error: $error"
            }
            .build()

        metronome = newMetronome

        setupObservers(newMetronome)
    }

    private fun setupObservers(metronome: SonixMetronome) {
        observerJobs.add(viewModelScope.launch {
            metronome.currentBeat.collect { beat ->
                _currentBeat.value = beat
            }
        })

        observerJobs.add(viewModelScope.launch {
            metronome.isPlaying.collect { playing ->
                _isRunning.value = playing
            }
        })

        observerJobs.add(viewModelScope.launch {
            metronome.isInitialized.collect { initialized ->
                _isInitialized.value = initialized
                if (initialized) {
                    _status.value = "Initialized"
                }
            }
        })

        observerJobs.add(viewModelScope.launch {
            metronome.error.collect { error ->
                error?.let {
                    _status.value = "Error: ${it.message}"
                }
            }
        })
    }

    fun start() {
        metronome?.start()
        _status.value = "Running"
    }

    fun stop() {
        metronome?.stop()
        _status.value = "Stopped"
        _currentBeat.value = 0
    }

    fun setBpm(newBpm: Float) {
        _bpm.value = newBpm
        metronome?.setBpm(newBpm)
    }

    fun setVolume(newVolume: Float) {
        _volume.value = newVolume
        metronome?.let { it.volume = newVolume }
    }

    fun setBeatsPerCycle(count: Int) {
        if (_beatsPerCycle.value == count) return
        _beatsPerCycle.value = count

        val samaPath = samaSamplePath
        val beatPath = beatSamplePath
        if (samaPath != null && beatPath != null) {
            createMetronome(samaPath, beatPath)
        }
    }

    override fun onCleared() {
        super.onCleared()
        observerJobs.forEach { it.cancel() }
        metronome?.release()
    }

    private fun copyAssetToFile(context: Context, assetName: String): File {
        val file = File(context.cacheDir, assetName)
        if (!file.exists()) {
            context.assets.open(assetName).use { input ->
                FileOutputStream(file).use { output ->
                    input.copyTo(output)
                }
            }
        }
        return file
    }
}

File: demo-apps/ios/VoxaTraceDemo/VoxaTraceDemo/Sections/Metronome/MetronomeViewModel.swift
import Foundation
import Combine
import VoxaTrace

/// ViewModel for metronome using SonixMetronome.Builder.
///
/// ## VoxaTrace Integration (~20 lines)
/// ```swift
/// // 1. Create metronome with Builder pattern
/// metronome = SonixMetronome.Builder()
///     .samaSamplePath(path: samaPath)
///     .beatSamplePath(path: beatPath)
///     .bpm(bpm: bpm)
///     .beatsPerCycle(count: beats)
///     .volume(volume: volume)
///     .build()
///
/// // 2. Observe state
/// metronome.observeIsInitialized { }
/// metronome.observeCurrentBeat { }
/// metronome.observeIsPlaying { }
///
/// // 3. Control
/// metronome.start() / metronome.stop()
/// metronome.setBpm(bpm:) / metronome.volume = newVolume
/// ```
@MainActor
final class MetronomeViewModel: ObservableObject {

    // MARK: - Published State

    @Published private(set) var isRunning = false
    @Published private(set) var isInitialized = false
    @Published private(set) var currentBeat: Int32 = 0
    @Published private(set) var status = "Ready"
    @Published var bpm: Float = 120
    @Published var beatsPerCycle: Int32 = 4
    @Published var volume: Float = 0.8

    // MARK: - Private

    private var metronome: SonixMetronome?
    private var beatTask: Task<Void, Never>?
    private var playingTask: Task<Void, Never>?
    private var initializedTask: Task<Void, Never>?
    private var errorTask: Task<Void, Never>?
    private var samaSamplePath: String?
    private var beatSamplePath: String?

    // MARK: - Lifecycle

    func onAppear() async {
        await initializeMetronome()
    }

    func onDisappear() {
        cancelObservers()
        metronome?.release()
    }

    // MARK: - Actions

    func start() {
        metronome?.start()
        status = "Running"
    }

    func stop() {
        metronome?.stop()
        status = "Stopped"
        currentBeat = 0
    }

    func setBpm(_ newBpm: Float) {
        bpm = newBpm
        metronome?.setBpm(bpm: newBpm)
    }

    func setVolume(_ newVolume: Float) {
        volume = newVolume
        metronome?.volume = newVolume
    }

    func setBeatsPerCycle(_ newBeats: Int32) {
        guard newBeats != beatsPerCycle else { return }
        beatsPerCycle = newBeats
        reinitializeMetronome(withBeats: newBeats)
    }

    // MARK: - Private Methods

    private func initializeMetronome() async {
        guard let samaPath = copyAssetToFile(name: "sama_click", ext: "wav"),
              let beatPath = copyAssetToFile(name: "beat_click", ext: "wav") else {
            status = "Click samples not found"
            return
        }

        samaSamplePath = samaPath
        beatSamplePath = beatPath

        // Create metronome with Builder pattern
        let newMetronome = SonixMetronome.Builder()
            .samaSamplePath(path: samaPath)
            .beatSamplePath(path: beatPath)
            .bpm(bpm: bpm)
            .beatsPerCycle(count: beatsPerCycle)
            .volume(volume: volume)
            .build()

        metronome = newMetronome
        status = "Initializing audio..."

        setupObservers(newMetronome)
    }

    private func reinitializeMetronome(withBeats newBeats: Int32) {
        guard let samaPath = samaSamplePath,
              let beatPath = beatSamplePath else {
            return
        }

        Task {
            cancelObservers()
            metronome?.release()

            isInitialized = false
            status = "Reinitializing..."

            let newMetronome = SonixMetronome.Builder()
                .samaSamplePath(path: samaPath)
                .beatSamplePath(path: beatPath)
                .bpm(bpm: bpm)
                .beatsPerCycle(count: newBeats)
                .volume(volume: volume)
                .build()

            metronome = newMetronome
            currentBeat = 0

            setupObservers(newMetronome)
        }
    }

    private func setupObservers(_ metronome: SonixMetronome) {
        initializedTask = metronome.observeIsInitialized { [weak self] initialized in
            self?.isInitialized = initialized
            if initialized {
                self?.status = "Ready"
            }
        }
        beatTask = metronome.observeCurrentBeat { [weak self] beat in
            self?.currentBeat = beat
        }
        playingTask = metronome.observeIsPlaying { [weak self] playing in
            self?.isRunning = playing
        }
        errorTask = metronome.observeError { [weak self] error in
            if let err = error {
                self?.status = "Error: \(err.message)"
            }
        }
    }

    private func cancelObservers() {
        beatTask?.cancel()
        playingTask?.cancel()
        initializedTask?.cancel()
        errorTask?.cancel()
    }
}
